Sources/MapboxDirections/Directions.swift

```diff
@@ -153,59 +116,85 @@ open class Directions: NSObject {
      - returns: The data task used to perform the HTTP request. If, while waiting for the completion handler to execute, you no longer want the resulting routes, cancel this task.
      */
     @discardableResult open func calculate(_ options: RouteOptions, completionHandler: @escaping RouteCompletionHandler) -> URLSessionDataTask {
-        let fetchStart = Date()
+        options.fetchStartDate = Date()
         let request = urlRequest(forCalculating: options)
         let requestTask = URLSession.shared.dataTask(with: request) { (possibleData, possibleResponse, possibleError) in
-            let responseEndDate = Date()
-            guard let response = possibleResponse, ["application/json", "text/html"].contains(response.mimeType) else {
+            
+            if let urlError = possibleError as? URLError {
+                let response = RouteResponse(httpResponse: possibleResponse as? HTTPURLResponse, options: .route(options), credentials: self.credentials)
                 DispatchQueue.main.async {
-                    completionHandler(nil, nil, .invalidResponse)
+                    completionHandler(response, .network(urlError))
+                }
+                return
+            }
+            
+            guard let response = possibleResponse, ["application/json", "text/html"].contains(response.mimeType), let httpResponse = response as? HTTPURLResponse else {
+                let response = RouteResponse(httpResponse: possibleResponse as? HTTPURLResponse, options: .route(options), credentials: self.credentials)
+                DispatchQueue.main.async {
+                    completionHandler(response, .invalidResponse(possibleResponse))
                 }
                 return
             }
             
             guard let data = possibleData else {
+                let response = RouteResponse(httpResponse: httpResponse, options: .route(options), credentials: self.credentials)
                 DispatchQueue.main.async {
-                    completionHandler(nil, nil, .noData)
+                    completionHandler(response, .noData)
                 }
                 return
             }
             
             if let error = possibleError {
+                let response = RouteResponse(httpResponse: httpResponse, options: .route(options), credentials: self.credentials)
+                let unknownError = DirectionsError.unknown(response: possibleResponse, underlying: error, code: nil, message: nil)
                 DispatchQueue.main.async {
-                    completionHandler(nil, nil, .unknown(response: possibleResponse, underlying: error, code: nil, message: nil))
+                    completionHandler(response, unknownError)
                 }
                 return
             }
```

Comment:
This check is probably not needed anymore. I think URLSession can only return URLError instances, so first check at the top is enough. Also this is dead code path. According to docs, `possibleError` is not nil only when `possibleData` is nil (and vice versa), so previous guard will enter its `else` clause before this if will be evaluated.

Reply:
👋 Hey @SebastianOsinski 

Great catch! yeah, that's definitely dead code. Will remove.